### ISSUE TITLE
DDP training problem

### ISSUE THREAD

**Issue by marson666** [webdataset/webdataset#225]
Hi, how to make sure the WebLoader generated batches is different data when use multi node training? Is there any up-to-date example code using webdataset for DDP training?

**Comment by senarvi** [comment]
Based on [this documentation](https://webdataset.github.io/webdataset/multinode/#splitting-shards-across-nodes-and-workers), you should create shards of the data and the library will by default distribute the shards to different nodes and workers. However, when using DDP, all processes need to receive the same number of mini-batches or they won't be able to synchronize. [The next section](https://webdataset.github.io/webdataset/multinode/#distributeddataparallel) says that you should use `ddp_equalize()` for this purpose, providing the total number of mini-batches, but according to [this other issue](https://github.com/webdataset/webdataset/issues/194) that functionality is broken and it might be better to switch to TorchData.

**Comment by senarvi** [comment]
> * You can choose not to split shards but just select every nth sample on each of the n nodes.

@tmbdev in this case, unless the data size is divisible by n, some nodes may get one more batch than the other nodes. How would you make sure that the synchronization won't hang, when some processes think that the epoch has finished already?

**Comment by senarvi** [comment]
This is what I ended up doing:

During training I cycle through the data indefinitely and set an arbitrary epoch length. The epoch length doesn't matter much, but it may affect the learning rate schedule.

During validation, I don't want to use an arbitrary epoch length, as that would mean that I either don't use the whole validation set, or some samples are tested multiple times. In this case I don't split the shards, but every GPU takes every Nth sample. Now I have the problem that if the validation set size is not divisible by N, those GPUs that receive an extra sample will be stuck waiting for the other GPUs, in the end of a validation epoch. I solved this by dropping samples from the end of the validation set to make the size divisible by `world_size`. This is what I used in TorchData (I didn't check how to do this using WebDataset):

```python
dataset = dataset.batch(torch.distributed.get_world_size(), drop_last=True)
dataset = dataset.unbatch()
dataset = dataset.sharding_filter()
```

**Comment by maxwellzh** [comment]
I'm also wondering this issue. 

For a more complicated case, there's usually a tar file containing less items (say, sequences) than others when using `wds.ShardWriter()` with a specific `maxcount`.

According to my understanding, solutions @tmbdev gives above just deal with uneven data on tar file level (that means every node would iter through the same number of tar files). But I think the DDP would hang if there is a tar of small size.

e.g.
```
data:  # 4100 seqs in total
data-001.tar # 2000 seqs
data-002.tar # 2000 seqs
data-003.tar # 100 seqs

split to 2 nodes (with duplicated)
node 0: data-001.tar data-003.tar # 2100 seqs
node 1: data-002.tar data-001.tar # 4000 seqs

2100 != 4000
```

Should I just drop the small size tar or do a post-processing after sharding to ensure all sharded tars are of close size.

**Comment by senarvi** [comment]
@maxwellzh no, the solutions work also when one tar file is smaller. If you use `resampled=True`, all processes will generate an infinite stream of samples, so DDP will never hang.

**Comment by rmchurch** [comment]
I did some testing with `resampled=True`, I was surprised to find that for the number of batches that just cover the number of total samples in my sharded dataset, only 64% of the samples were unique. For how the data is setup, I was expecting <<1% to be duplicate data in order to create even batch sizes. Why is this? Is there a way to only resample for the last samples which are needed? It appears it is randomly sampling, giving a lot of overlap (and leaving out a lot of data).

**Comment by nicolas-dufour** [comment]
Hi @tmbdev, i'm trying to use your implementation with .repeat(2).with_epoch(n), but i'm not sure what does repeat is supposed to do. Is 2 the number of gpus per node? The number of nodes? the overall number of gpus?

**Comment by senarvi** [comment]
> Hi @tmbdev, i'm trying to use your implementation with .repeat(2).with_epoch(n), but i'm not sure what does repeat is supposed to do. Is 2 the number of gpus per node? The number of nodes? the overall number of gpus?

Iterates through the data twice, so that you won't run out of data before you've reached n samples.

**Comment by nicolas-dufour** [comment]
ok thanks

**Comment by tmbdev** [comment]
Resampling is quite simple: it just picks shards at random within each worker. The number of unique samples you get per epoch depends on the number of shards you have and the size of your epoch. If you have few shards, there is a high probability that different workers will pick the same shard. You have many different ways of modifying this strategy. For example, you can split shards by node and worker prior to resampling. But in the end, you really need a number of shards that is much larger than the number of total workers in your dataset to avoid significant numbers of duplicate samples with resampling.

The best thing to do is to make a number of shards that is divisible by the total number of workers in your training job and have exactly the same number of samples in each shard. Then you can assign each worker the same number of shards. In that case, you get exact epochs with no resampling, no duplication, and no missing samples. But obviously that requires more setup.

You also have the option of using various forms of async training with PyTorch, so that you can do weight updates even if not all workers have data.

There is only a limited set of things we can do within the old torch data architecture, since workers cannot communicate between themselves. Torchdata will provide some workarounds.
